### ISSUE TITLE
Input fields in the table losing focus after typing

### ISSUE THREAD

**Issue by canercak** [jbetancur/react-data-table-component#571]
Thank you for this great library. I have a dynamic list with inputs as seen on the picture. Every time I try to write text in a field it loses focus so I cannot write.  I tried a lot including placing autofocus, adding keys to divs and different kinds of input fields. The same thing happens if I use simple variables(variantList) rather than field values(values.variant_list) as well. The same code works with reactstrap table, material-table and if I use html table.   
```

   const variantTableColumns = [
    {
      name: t('product.field.variant_name'),
      sortable: true,
      cell: row => <div>{row.name}</div>
    },
    {
      name: t('product.field.sku'),
      sortable: true,
      cell: row => (
         <div key={`variantlist1-${row.idx}`}  >
          <Input
            name={`sku-${row.idx}`}
            id={`sku-${row.idx}`}
            key={`sku-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'sku')}
            type="text"
            value={values.variant_list[row.idx].sku}
          />
        </div>
      )
    },
    {
      name: t('product.field.barcode'),
      sortable: true,
      cell: row => (
        <div key={`variantlist2-${row.idx}`}  >
          <Input
            name={`barcode-${row.idx}`}
            id={`barcode-${row.idx}`}
            key={`barcode-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'barcode')}
            type="text"
            value={values.variant_list[row.idx].barcode}
          />
        </div>
      )
    },
    {
      name: t('product.field.supplier_code'),
      sortable: true,
      cell: row => (
        <div key={`variantlist3-${row.idx}`}  >
          <Input
            name={`supplier_code-${row.idx}`}
            id={`supplier_code-${row.idx}`}
            key={`supplier_code-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'supplier_code')}
            type="text"
            value={values.variant_list[row.idx].supplier_code}
          />
        </div>
      )
    },
    {
      name: t('product.field.supply_price'),
      sortable: true,
      cell: row => (
        <div key={`variantlist4-${row.idx}`}  >
          <Input
            name={`supply_price-${row.idx}`}
            id={`supply_price-${row.idx}`}
            key={`supply_price-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'supply_price')}
            type="text"
            value={values.variant_list[row.idx].supply_price}
          />
        </div>
      )
    },
    {
      name: t('product.field.retail_price'),
      sortable: true,
      cell: row => (
        <div key={`variantlist5-${row.idx}`}  >
          <Input
            name={`retail_price-${row.idx}`}
            id={`retail_price-${row.idx}`}
            key={`retail_price-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'retail_price')}
            type="text"
            value={values.variant_list[row.idx].retail_price}
          />
        </div>
      )
    },
    {
      name: t('product.field.variant_active'),
      sortable: true,
      cell: row => (
        <div key={`variantlist6-${row.idx}`}  >
          <Switch
            name={`active-${row.idx}`}
            id={`active-${row.idx}`}
            key={`active-${row.idx}`}
            data-idx={row.idx}
            onChange={e=>handleVariantListItemChange(e,'active', row.idx)}
            checked={values.variant_list[row.idx].active}
            className="react-switch-block"
          />
        </div>
      )
    },
    {
      sortable: false,
      right: true,
      cell: row => (
        <div>
          <Button sm color="secondary" class="btn-sm" onClick={event => handleRemoveVariantListItem(event,row.id)}>
            <FontAwesomeIcon icon={faTrashAlt} style={{ fontSize: 14,  marginBottom: '1px' }} />
          </Button>
        </div>
      )
    }
  ];
```

```
  const handleVariantListItemChange = (e,classname, idx=null) => {
    let listvariants = values.variant_list;
    if (classname === 'active') {
      listvariants[idx][classname] = e
    } else {
      listvariants[e.target.dataset.idx][classname] = e.target.value;
    }
    setFieldValue('variant_list', listvariants);
  };
```
```

 <DataTable
              columns={variantTableColumns}
              customStyles={variantTableStyles}
              striped={true}
              striped={true}
              fixedHeader={true}
              data={values.variant_list}
            />
```
![Screenshot at May 09 06-03-10](https://user-images.githubusercontent.com/2890782/81462409-ecbd6080-91ba-11ea-8bca-ca965cae45ab.png)

**Comment by slcastroj** [comment]
This specific issue is so horrible for my project, I need to update a column given 2 inputs in the table and it just loses focus so I can't type more than 1 character.

Of course, the issue is caused because it is re-rendering, and I'm pretty sure this could be worked around if the cell keys where static, but I might be wrong here.

**Comment by canercak** [comment]
This is a real issue. I tried onBlur without success as well. I'm using react material-table now, it works fine.

**Comment by alexpo27** [comment]
Did anyone find a fix for the issue? I maanged to make it work by adding:
`            setTimeout(() => {
                document.getElementById('client-table-search').focus();
            }, 0)`
Not the most elegant fix but it does the job for now.

**Comment by Clemios** [comment]
I'm facing the same issue here
![Screenshot 2020-05-25 at 17 27 56](https://user-images.githubusercontent.com/9378934/82826380-4d33e980-9ead-11ea-86f2-7a896460d5d5.png)

I'm using `react-data-table-component` with a switch to pass the line to edit mode with inputs instead of text.

I've tried to move the state management to the right place to avoid an entire re-rendering and stoping loosing focus.

I've tried to define `Columns` with state management, i'v tried at the index.js level, i also tried at the cell component level.

I have also tried to add keys, to programaticaly set the focus with a useFocus hook, after many tries i wasn't able to do it.

**Comment by Clemios** [comment]
I've done this codesandbox to reproduce the issue. I tried to use `React.useMemo` to solve the problem with no success :(

https://codesandbox.io/s/react-data-table-materialui-k2043?file=/src/index.js

**Comment by Clemios** [comment]
> Did anyone find a fix for the issue? I maanged to make it work by adding:
> ` setTimeout(() => { document.getElementById('client-table-search').focus(); }, 0)`
> Not the most elegant fix but it does the job for now.

Thank you for the trick, i have implemented an exemple of your solution here : 
https://codesandbox.io/s/react-data-table-materialui-ocu46?file=/src/index.js

It makes the input focus indicator flash at each update but its working. I'm quite sure that in certain case the state update could be anormaly long and it can result as a loss of caracters if you type quickly.

**Comment by kiriman** [comment]
One of ways, your can set "autofocus" attribute: <input ... autofocus />

**Comment by canercak** [comment]
@Clemios thank you, timeout works. Autofocus doesn't work for me.

**Comment by nimesh5692** [comment]
> 
> 
> One of ways, your can set "autofocus" attribute: <input ... autofocus />

@kiriman input blinks like hell, but it works. Thanks 😁

**Comment by tpham** [comment]
This is a real issue and I hope it will get addressed in the component itself.

**Comment by jbetancur** [comment]
I’ll have a look at this as soon as I am able

**Comment by jbetancur** [comment]
Sorry for not seeing this issue earlier. With everything going on and increased interest in RDT, it's been a challenge keeping up.

I've been in the process of refactoring to use Typescript. During this "journey", I've cleaned up some issues pertaining to React context and how things were rendering. 

I tried this forked sandbox above posted by @Clemios and I changed the RDT version from 6.3.1 to 7.0.0-alpha-4 and I no longer can dupe the issue.

https://codesandbox.io/s/react-data-table-materialui-forked-x53ec

TBH I have no idea what changes during the refactor fixed this, but it seems to be working.

I know that 7.0.0 is an alpha version, but there are only a couple of breaking changes in 7.0.0 with `defaultSortfField` which is now `defaultSortFieldId` where you use a column ID instead of a name, but the API remains the same for now.
